tests/func/test_diff.py

```diff
@@ -264,3 +264,148 @@ def test_no_commits(tmp_dir):
     assert Git().no_commits
 
     assert Repo.init().diff() == {}
+
+
+def setup_targets_test(tmp_dir):
+    tmp_dir.dvc_gen("file", "first", commit="add a file")
+
+    tmp_dir.dvc_gen({"dir": {"1": "1", "2": "2"}})
+    tmp_dir.dvc_gen("file", "second")
+
+    tmp_dir.dvc_gen(os.path.join("dir_with", "file.txt"), "first")
+
+    remove(tmp_dir / ".dvc" / "cache")
+    (tmp_dir / ".dvc" / "tmp" / "state").unlink()
+
+
+def test_targets_missing_path(tmp_dir, scm, dvc):
+    from dvc.exceptions import PathMissingError
+
+    setup_targets_test(tmp_dir)
+
+    with pytest.raises(PathMissingError):
+        dvc.diff(targets=["missing"])
+
+
+def test_targets_single_file(tmp_dir, scm, dvc):
+    setup_targets_test(tmp_dir)
+
+    assert dvc.diff(targets=["file"]) == {
+        "added": [],
+        "deleted": [],
+        "modified": [
+            {
+                "path": "file",
+                "hash": {"old": digest("first"), "new": digest("second")},
+            }
+        ],
+        "not in cache": [],
+    }
+
+
+def test_targets_single_dir(tmp_dir, scm, dvc):
+    setup_targets_test(tmp_dir)
+
+    dir_checksum = "5fb6b29836c388e093ca0715c872fe2a.dir"
+
+    expected_result = {
+        "added": [
+            {"path": os.path.join("dir", ""), "hash": dir_checksum},
+            {"path": os.path.join("dir", "1"), "hash": digest("1")},
+            {"path": os.path.join("dir", "2"), "hash": digest("2")},
+        ],
+        "deleted": [],
+        "modified": [],
+        "not in cache": [],
+    }
+
+    assert dvc.diff(targets=["dir"]) == expected_result
+    assert dvc.diff(targets=["dir" + os.path.sep]) == expected_result
+
+
+def test_targets_single_file_in_dir(tmp_dir, scm, dvc):
+    setup_targets_test(tmp_dir)
+
+    dvc.diff(targets=["dir" + os.path.sep])
```

Comment:
@efiop @pmrowla There's some weirdness here if this line is removed, an error is thrown:

```
E       dvc.config.NoRemoteError: config file error: no remote specified. Create a default remote with
E           dvc remote add -d <remote name> <remote url>
```

I'm not sure if `setup_targets_test` is causing this or if an actual issue in dvc.

Reply:
It's because you are deleting the cache directory in `setup_targets_test`. We delete the cache directory in some of the existing diff tests because we are specifically testing for the `cache missing` diff state there. For these `targets` related tests you should not need to delete cache or unlink state in your setup function.

Reply:
Thanks! Updated in https://github.com/iterative/dvc/pull/4938/commits/1b6ee154c09370f4cd16aa2b58c5a457850a3259.